go/vt/vtgate/planbuilder/operators/plan_query.go

```diff
@@ -73,13 +74,28 @@ func PlanQuery(ctx *plancontext.PlanningContext, stmt sqlparser.Statement) (resu
 	checkValid(op)
 	op = planQuery(ctx, op)
 
-	_, isRoute := op.(*Route)
-	if !isRoute && ctx.SemTable.NotSingleRouteErr != nil {
-		// If we got here, we don't have a single shard plan
-		return nil, ctx.SemTable.NotSingleRouteErr
+	err = shouldErrorOnComplexPlan(ctx, op)
+	return op, err
+}
+
+func shouldErrorOnComplexPlan(ctx *plancontext.PlanningContext, op Operator) error {
+	complexPlan := false
+	visitF := func(op Operator, _ semantics.TableSet, _ bool) (Operator, *ApplyResult) {
+		switch op.(type) {
+		case *Limit, *Route:
+		default:
+			complexPlan = true
+		}
+		return op, NoRewrite
 	}
 
-	return op, err
+	TopDown(op, TableID, visitF, stopAtRoute)
+
+	if complexPlan && ctx.SemTable.NotSingleRouteErr != nil {
```

Comment:
we could check `ctx.SemTable.NotSingleRouteErr` before checking if we are dealing with a complex plan. if it's nil, no need to walk the operator tree.